libcontainer/nsenter/nsexec.c

```diff
@@ -242,19 +320,75 @@ void nsexec()
 	// We must fork to actually enter the PID namespace, use CLONE_PARENT
 	// so the child can have the right parent, and we don't need to forward
 	// the child's exit code or resend its death signal.
-	child = clone_parent(&env);
+	int child = clone_parent(&env, cloneflags);
 	if (child < 0) {
 		pr_perror("Unable to fork");
 		exit(1);
 	}
+	// if uid_map and gid_map were specified, writes the data to /proc files
+	if (uidmap_start > 0 && uidmap_len > 0) {
+		char buf[PATH_MAX];
+		if (snprintf(buf, sizeof(buf), "/proc/%d/uid_map", child) < 0) {
+			pr_perror("failed to construct uid_map file for %d",
+				  child);
+			exit(1);
+		}
 
-	len = snprintf(buf, sizeof(buf), "{ \"pid\" : %d }\n", child);
+		int fd = open(buf, O_RDWR);
+		writedata(fd, data, uidmap_start, uidmap_len);
+	}
 
-	if (write(pipenum, buf, len) != len) {
+	if (gidmap_start > 0 && gidmap_len > 0) {
+		if (is_setgroup == 1) {
+			char buf[PATH_MAX];
+			if (snprintf
+			    (buf, sizeof(buf), "/proc/%d/setgroups",
+			     child) < 0) {
+				pr_perror
+				    ("failed to construct setgroups file for %d",
+				     child);
+				exit(1);
+			}
+
+			int fd = open(buf, O_RDWR);
+			if (write(fd, "allow", 5) != 5) {
+				// If the kernel is too old to support /proc/PID/setgroups,
+				// write will return ENOENT; this is OK.
+				if (errno != ENOENT) {
+					pr_perror("failed to write allow to %s",
+						  buf);
+					exit(1);
+				}
+			}
+		}
+		// write gid mappings
+		char buf[PATH_MAX];
+		if (snprintf(buf, sizeof(buf), "/proc/%d/gid_map", child) < 0) {
+			pr_perror("failed to construct gid_map file for %d",
+				  child);
+			exit(1);
+		}
+
+		int fd = open(buf, O_RDWR);
+		writedata(fd, data, gidmap_start, gidmap_len);
```

Comment:
You break the coding style for this file. For me, it's not obvious why we don't check exit codes for open() and writedata(). I would prefer to declare all variables at the beginning of blocks